pkg/kubelet/kuberuntime/kuberuntime_manager.go

```diff
@@ -447,19 +457,35 @@ func (m *kubeGenericRuntimeManager) computePodContainerChanges(pod *v1.Pod, podS
 	// check the status of containers.
 	for index, container := range pod.Spec.Containers {
 		containerStatus := podStatus.FindContainerStatusByName(container.Name)
-		if containerStatus == nil || containerStatus.State != kubecontainer.ContainerStateRunning {
+		containerRef, err := kubecontainer.GenerateContainerRef(pod, &container)
+		if err != nil {
+			glog.Errorf("Couldn't make a ref to pod %q, container %v: '%v'", format.Pod(pod), container.Name, err)
+		}
+
+		if containerStatus == nil {
+			// No container exists by this name, so just start it
+			changes.ContainersToStart[index] = "Starting container"
+			continue
+		} else if containerStatus.State != kubecontainer.ContainerStateRunning {
+			// The container exists, but isn't running.  It must have exited.
+			exitMessage := fmt.Sprintf("Container exited with status %d", containerStatus.ExitCode)
+			if containerStatus.ExitCode == 0 {
+				m.recorder.Eventf(containerRef, v1.EventTypeNormal, events.ExitSuccess, exitMessage)
+			} else {
+				m.recorder.Eventf(containerRef, v1.EventTypeWarning, events.ExitFailure, exitMessage)
```

Comment:
Ok this is a little harder than I thought, since there's no "previous state" in scope at this point.

I'm exploring plumbing `existingStatus` in from kubelet.go:1389 through to SyncPod, which seems to work well.  I'll do some more testing and amend.